sk_sidekick/lib/src/commands/lock_dependencies_command.dart

```diff
@@ -0,0 +1,102 @@
+import 'package:sidekick_core/sidekick_core.dart';
+import 'package:sk_sidekick/src/commands/bump_version_command.dart';
+import 'package:yaml/yaml.dart';
+
+class LockDependenciesCommand extends Command {
+  LockDependenciesCommand() {}
+
+  @override
+  final String description = 'Locks all dependencies to their upper bounds';
+
+  @override
+  final String name = 'lock-dependencies';
+
+  @override
+  String get invocation => super.invocation.replaceFirst(
+        '[arguments]',
+        '[package-path]',
+      );
+
+  @override
+  Future<void> run() async {
+    final packagePath = argResults!.rest.firstOrNull ?? Directory.current.path;
+    final package = DartPackage.fromDirectory(Directory(packagePath));
+    if (package == null) {
+      throw 'Could not find a package in $packagePath';
+    }
+
+    final pubGet = systemDart(['pub', 'get'], workingDirectory: package.root);
+    if (pubGet != 0) {
+      throw "Couldn't get dependencies in ${package.root}";
+    }
+
+    final lockfile = package.lockfile;
+    if (!lockfile.existsSync()) {
+      throw "Lockfile doesn't exist in ${package.root}";
+    }
+
+    final packages =
+        loadYaml(lockfile.readAsStringSync())['packages'] as YamlMap;
+
+    final Map<String, String> directDependencies = {};
+    final Map<String, String> transitiveDependencies = {};
+    for (final package in packages.entries) {
+      final type = package.value['dependency'];
+      final packageName = package.key as String;
+      final version = package.value['version'] as String;
+
+      if (type != 'direct dev' && package.value['source'] != 'hosted') {
+        throw "Can only handle dependencies from hosted sources, but $packageName violates this.";
+      }
+
+      switch (type) {
+        case 'transitive':
+          transitiveDependencies[packageName] = version;
+          break;
+        case 'direct main':
+          directDependencies[packageName] = version;
+          break;
+        // case 'direct dev' is irrelevant
+      }
+    }
+
+    final pinnedDirectDependencies = directDependencies
+        .mapEntries<String>((e) => "  ${e.key}: ${e.value.lockedConstraint}")
+        .sorted();
+    final pinnedTransitiveDependencies = transitiveDependencies
+        .mapEntries<String>((e) => "  ${e.key}: ${e.value.lockedConstraint}")
+        .sorted();
```

Comment:
This writes the exact versions of what's in the lock file into the `pubspec.yaml`. That doesn't work because we have support **different dart versions**.
i.e. `win32:2.6.0` (which would be locked) when running with dart 2.17+ is incompatible with dart 2.16 and below. Those older dart versions require `win32:2.5.2` (dart 2.15) or `win32:2.3.6` (dart 2.14).

We have to lock a range. We should take the minimum version as defined in the current `pubspec.yaml` and get the latest available version from pub.dev and lock that as upper constraint.

Reply:
works as expected 👍